### PR TITLE
More custom VEP annotation types

### PR DESCRIPTION
**Pull request by tedil**
For some of the annotations I had to guess their possible values and separators, since neither
https://www.ensembl.org/info/docs/tools/vep/online/input.html#ident
nor
https://www.ensembl.org/info/docs/tools/vep/vep_formats.html
nor
https://www.ensembl.org/Help/Glossary?id=346
has any information about them. But I think we're good.

### CODE REVIEW

File: vembrane/ann_types.py [hunk 1]
```diff
@@ -140,12 +141,50 @@ def description(self):
...
+    def __init__(self, name: str, **kwargs):
+        def typefunc(x):
+            key_value_tuples = map(lambda v: v.split(":", maxsplit=1), x.split("&"))
+            mapping = defaultdict(list)
+            for key, value in key_value_tuples:
+                mapping[key].append(value)
+            # make sure non-existent keys do not return list() on __getitem__ calls
+            # in the filter expression (without actually building a new dict from
+            # the defaultdict)
+            mapping.default_factory = None
```
Comment (mbargull) [inline vembrane/ann_types.py:169]: So, it is intentional that a user lookup returns `None` for unknown values? (Just asking -- I didn't look at the surrounding code at all. I.e., I also don't know whether we want `None` or `NA` here.)

File: vembrane/ann_types.py [hunk 2]
```diff
@@ -140,12 +141,50 @@ def description(self):
...
+
+class AnnotationNumberTotalEntry(AnnotationListEntry):
+    def __init__(self, name: str, **kwargs):
+        super().__init__(name, sep="/", inner_typefunc=int, **kwargs)
+
+
+class AnnotationPredictionScoreEntry(AnnotationEntry):
+    def __init__(self, name: str, **kwargs):
+        def typefunc(x: str) -> Dict[str, float]:
+            return {x[: x.rindex("(")]: float(x[x.index("(") + 1 : x.rindex(")")])}
```
Comment (mbargull) [inline vembrane/ann_types.py:183]:
```suggestion
            return {x[: x.index("(")]: float(x[x.index("(") + 1 : x.rindex(")")])}
```
(Or use a regex. Regex is probably a little more readable, but I'm okay with both.)

File: vembrane/ann_types.py [hunk 3]
```diff
@@ -289,14 +324,12 @@ def convert(self, key: str, value: str) -> Tuple[str, AnnotationType]:
...
-    #  "a flag indicating if the variant falls in a high information position of a
-    #  transcription factor binding profile (TFBP)"
-    # "HIGH_INF_POS": AnnotationEntry(
-    #     "HIGH_INF_POS",
-    #     description="A flag indicating if the variant falls in a high information "
-    #                 "position of a transcription factor binding profile (TFBP)",
-    # ),
+    "HIGH_INF_POS": AnnotationEntry(
+        "HIGH_INF_POS",
+        typefunc=lambda x: True if x == "Y" else False,
```
Comment (mbargull) [inline vembrane/ann_types.py:329]:
```suggestion
        typefunc=lambda x: if x == "Y",
```
Though, if you think `True if ... else False` is more readable, I'm happy to leave it as is!
If micro-optimization performance is of interest,
```suggestion
        typefunc="Y".__eq__,
```
;).

File: vembrane/ann_types.py [hunk 4]
```diff
@@ -308,29 +341,25 @@ def convert(self, key: str, value: str) -> Tuple[str, AnnotationType]:
...
-    #  "a flag indicating if the transcript is denoted as the canonical transcript
-    #  for this gene"
-    # "CANONICAL": AnnotationEntry(
-    #     "CANONICAL",
-    #     description="A flag indicating if the transcript is denoted as the canonical "
-    #                 "transcript for this gene",
-    # ),
+    "CANONICAL": AnnotationEntry(
+        "CANONICAL",
+        typefunc=lambda x: True if x == "YES" else False,
```
Comment (mbargull) [inline vembrane/ann_types.py:346]: (Same as above)
```suggestion
        typefunc=lambda x: if x == "YES",
```
Though, if you think `True if ... else False` is more readable, I'm happy to leave it as is!
If micro-optimization performance is of interest,
```suggestion
        typefunc="YES".__eq__,
```
;).

File: vembrane/common.py [hunk 5]
```diff
@@ -32,7 +32,10 @@ def get_annotation_keys(header: VariantHeader, ann_key: str) -> List[str]:
             continue
         if rec.get("ID") == ann_key:
             return list(
-                map(str.strip, rec.get("Description").split(separator)[1].split("|"))
+                map(
+                    str.strip,
+                    rec.get("Description").strip('"').split(separator)[1].split("|"),
```
Comment (mbargull) [inline vembrane/common.py:37]: Is the content of `Description` somewhat well-formed/restricted? Or could it also be something like
`"Description": '"Foo: Yadda yadda "quote""'` <- having an additional `"` before the end double quote.

File: vembrane/ann_types.py [hunk 6]
```diff
@@ -140,12 +141,50 @@ def description(self):
...
+
+        super().__init__(name=name, typefunc=typefunc, **kwargs)
+
+
+class AnnotationNumberTotalEntry(AnnotationListEntry):
+    def __init__(self, name: str, **kwargs):
+        super().__init__(name, sep="/", inner_typefunc=int, **kwargs)
+
+
+class AnnotationPredictionScoreEntry(AnnotationEntry):
```
Comment (mbargull) [inline vembrane/ann_types.py:180]: Is it intentional that these are lists of `int`/`float`? (Again, didn't look at the context here, sorry.)

File: vembrane/ann_types.py (hunk 1 above)
Comment (tedil) [reply vembrane/ann_types.py:169]: Oh well, good point. Probably something akin to the `nafunc` but with one (unneeded) param? Didn't think that one through

File: vembrane/ann_types.py (hunk 2 above)
Comment (tedil) [reply vembrane/ann_types.py:183]: ... and we could compile it in advance, might do that

File: vembrane/common.py (hunk 5 above)
Comment (tedil) [reply vembrane/common.py:37]: it's well formed, as far as I know.

File: vembrane/ann_types.py (hunk 6 above)
Comment (tedil) [reply vembrane/ann_types.py:180]: Yes, basically a pair `[number, total]`. Perhaps a custom type with number and total slots might be better?

File: vembrane/ann_types.py (hunk 2 above)
Comment (tedil) [reply vembrane/ann_types.py:183]: `(.*)\((.*)\)` should do the trick :D

Comment (tedil) [reply vembrane/ann_types.py]: but is it more readable... I don't know.

**Review by christopher-schroeder — APPROVED** [review]

**Pull request merged**